improver_tests/calibration/rainforests_calibration/test_ApplyRainForestsCalibration.py

```diff
@@ -115,3 +117,350 @@ def test__init__(
+        name="lwe_thickness_of_precipitation_amount",
+        units="m",
+        realizations=np.arange(10),
+    )
+    with pytest.raises(ValueError):
+        ApplyRainForestsCalibration(
+            model_config_dict={}, threads=1
+        )._align_feature_variables(ensemble_features, misaligned_forecast_cube)
+
+
+def test__prepare_error_probability_cube(
+    ensemble_forecast, error_thresholds, error_threshold_cube
+):
+    """Test the preparation of error probability cube from input
+    forecast cube."""
+    plugin = ApplyRainForestsCalibration(model_config_dict={}, threads=1)
+    plugin.error_thresholds = error_thresholds
+    result = plugin._prepare_error_probability_cube(ensemble_forecast)
+
+    assert result.long_name == error_threshold_cube.long_name
+    assert result.units == error_threshold_cube.units
+    assert result.coords() == error_threshold_cube.coords()
+    assert result.attributes == error_threshold_cube.attributes
+
+
+def test__prepare_features_dataframe(ensemble_features):
+    """Test dataframe preparation given set of feature cubes."""
+    feature_names = [cube.name() for cube in ensemble_features]
+    expected_size = ensemble_features.extract_cube(
+        "lwe_thickness_of_precipitation_amount"
+    ).data.size
+    result = ApplyRainForestsCalibration(
+        model_config_dict={}, threads=1
+    )._prepare_features_dataframe(ensemble_features)
+
+    assert list(result.columns) == list(sorted(feature_names))
+    assert len(result) == expected_size
+
+    # Drop realization coordinate from one of the ensemble features, to produce
+    # cubes of differing length.
+    cube_lacking_realization = ensemble_features.pop(-1).extract(
+        Constraint(realization=0)
+    )
+    cube_lacking_realization.remove_coord("realization")
+    ensemble_features.append(cube_lacking_realization)
+    with pytest.raises(RuntimeError):
+        ApplyRainForestsCalibration(
+            model_config_dict={}, threads=1
+        )._prepare_features_dataframe(ensemble_features)
+
+
+def test_make_decreasing():
+    """Test that make_increasing returns an array that is non-decreasing
+    in the first dimension."""
+    input_array = np.array([[5, 5], [4, 3], [3, 4], [2, 2], [1, 1]]) / 5.0
+    expected = np.array([[5, 5], [4, 3.5], [3, 3.5], [2, 2], [1, 1]]) / 5.0
+    result = ApplyRainForestsCalibration(
+        model_config_dict={}, threads=1
+    )._make_decreasing(input_array)
+    assert np.allclose(expected, result)
+
+
+def test__calculate_error_probabilities_lightgbm(
+    ensemble_features, ensemble_forecast, dummy_lightgbm_models
+):
+    """Test calculation of error probability cube when using lightgbm Boosters."""
+    plugin = ApplyRainForestsCalibration(model_config_dict={}, threads=1)
+    plugin.tree_models, plugin.error_thresholds = dummy_lightgbm_models
+    result = plugin._calculate_error_probabilities(ensemble_forecast, ensemble_features)
+
+    # Check that data has sensible probability values
```

Comment:
Good suggestion. This assumption was in the back of mind (so intentional as you highlight), however I agree that it is worth being explicit here.

This is tad verbose, but hopefully describes the intention of this test:
```suggestion
    # Check that data has sensible probability values.
    # Note: here we are NOT checking the returned value against an expected value
    # as this will be sensitive to changes in associated GBDT libraries, given that
    # the tree models are created dynamically within fixtures. Here we implicitly trust
    # the output from the tree models are correct based on the specified inputs, and so
    # only test to ensure that the dataset overall conforms to the bounds for probability
    # data.
```